### PR TITLE
add sql-bench test

### PR DESCRIPTION
**Pull request by cwen0**
* add bench test for special SQL

### CODE REVIEW

File: sql-bench/main.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+	addr           = flag.String("addr", ":4000", "tidb-server addr, default: :4000")
+	dbName         = flag.String("db", "test", "db name, default: test")
+	user           = flag.String("u", "root", "username, default: root")
+	password       = flag.String("p", "", "password, default: empty")
+	logLevel       = flag.String("L", "error", "log level, default: error")
+	sqlFile        = flag.String("data", "./bench.sql", "SQL data file for bench")
+)
+
+var (
+	connPool = pool.NewCache("pool", *poolSize, func() interface{} {
```
Comment (tennix) [inline sql-bench/main.go]: create connPool in init function

File: sql-bench/main.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+func init() {
+	flag.Parse()
+	statChan = make(chan *stat, 10000)
+}
+
+func cleanup() {
+	// Do nothing
+	for i := 0; i < *poolSize; i++ {
+		db := connPool.Get().(*sql.DB)
```
Comment (tennix) [inline sql-bench/main.go]: type assertion should be checked

File: sql-bench/main.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	}
+	execTime := time.Now().Sub(startTs)
+	fmt.Println("\n*************************final result***************************\n")
+	fmt.Printf("Total Query: %d, Succ: %d, Faild: %d, Time: %v, Avg response time: %.04fms, QPS: %.02f : \n", total, succ, total-succ, execTime, (spend.Seconds()*1000)/float64(total), float64(total)/execTime.Seconds())
+}
+
+func main() {
+	// Start
+	fmt.Println("Start Bench")
+	log.SetLevelByString(*logLevel)
```
Comment (tennix) [inline sql-bench/main.go]: set log level in `init` function

File: sql-bench/main.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+			log.Fatal("break done")
+		}
+		return db
+	})
+	statChan chan *stat
+)
+
+func init() {
+	flag.Parse()
+	statChan = make(chan *stat, 10000)
```
Comment (tennix) [inline sql-bench/main.go]: You'd better define`statChan` and `queryChan` size 10000 as `const`

File: sql-bench/parse/parse.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		log.Fatalf("Cteate out.sql Error: %s", err)
+	}
+	defer func() {
+		file.Close()
+	}()
+	dataArr := bytes.Split(data, []byte("Query"))
+	for _, v := range dataArr {
+		sql := strings.Replace(string(v), "\n", " ", -1)
+		sql = strings.Trim(strings.TrimSpace(sql), "\n")
```
Comment (tennix) [inline sql-bench/parse/parse.go]: `strings.TrimSpace` already removes `\n`

File: sql-bench/parse/parse.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		log.Fatalf("ReadFile Error: %s", err)
+	}
+	if len(data) == 0 {
+		log.Fatal("file is empty")
+	}
+	file, err := os.Create(*outFile)
+	if err != nil {
+		log.Fatalf("Cteate out.sql Error: %s", err)
+	}
+	defer func() {
```
Comment (tennix) [inline sql-bench/parse/parse.go]: `defer file.Close()`

File: sql-bench/main.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+	go statWorker(&wgStat, startTs)
+
+	ctxR, _ := context.WithTimeout(context.Background(), timeout)
+	go readQuery(ctxR, queryChan)
+	wg.Wait()
+	close(statChan)
+	cleanup()
+	wgStat.Wait()
+	fmt.Println("Done!")
+	return
```
Comment (tennix) [inline sql-bench/main.go]: `return` not needed

File: sql-bench/main.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	sql := strings.ToLower(sqlStmt)
+	isQuery := strings.HasPrefix(sql, "select")
+	// Get time
+	startTs := time.Now()
+	err := runQuery(sqlStmt, isQuery)
+	if err != nil {
+		log.Warnf("Exec sql [%s]: %s", sqlStmt, err)
+	}
+	spend := time.Now().Sub(startTs)
+	s := &stat{spend: spend, succ: err == nil}
```
Comment (tennix) [inline sql-bench/main.go]: when failed, the stat is useless, just pass an empty stat to `statChan`
```
if err != nil {
    log.Warnf()
    statChan <- &stat{}
    return err
}
statChan <- &stat{spend: time.Now().Sub(startTs), succ: true}
return nil
```

File: sql-bench/main.go [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+func runQuery(sqlStmt string, isQuery bool) error {
+	db := connPool.Get().(*sql.DB)
+	defer connPool.Put(db)
+	if isQuery {
+		rows, err := db.Query(sqlStmt)
+		if err != nil {
+			return err
+		}
+		for {
```
Comment (tennix) [inline sql-bench/main.go]: No need to retrieve rows

**Comment by tennix** [comment]
LGTM

File: sql-bench/main.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (andelf) [inline sql-bench/main.go]: lacks year

File: sql-bench/main.go [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+		return db
+	})
+}
+
+func cleanup() {
+	// Do nothing
+	for i := 0; i < *poolSize; i++ {
+		db, ok := connPool.Get().(*sql.DB)
+		if !ok {
+			log.Fatal("The type of db getted from pool is wrong")
```
Comment (andelf) [inline sql-bench/main.go]: s/getted/got/

File: sql-bench/main.go [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+		query := scanner.Text()
+		cnt++
+		queryChan <- query
+		querys = append(querys, query)
+		if *sqlCount != 0 && cnt >= *sqlCount {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			fmt.Printf("Get %d queries\n", cnt)
```
Comment (andelf) [inline sql-bench/main.go]: use log

File: sql-bench/main.go [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+	for cnt < *sqlCount {
+		for _, query := range querys {
+			cnt++
+			queryChan <- query
+			if cnt >= *sqlCount {
+				break LOOP
+			}
+			select {
+			case <-ctx.Done():
+				fmt.Printf("Get %d queries\n", cnt)
```
Comment (andelf) [inline sql-bench/main.go]: use log

File: sql-bench/main.go [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+				return
+			default:
+			}
+		}
+	}
+	fmt.Printf("Get %d queries\n", cnt)
```
Comment (andelf) [inline sql-bench/main.go]: ~

File: sql-bench/main.go [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+		return err
+	}
+	statChan <- &stat{spend: time.Now().Sub(startTs), succ: true}
+	return nil
+}
+
+func runQuery(sqlStmt string, isQuery bool) error {
+	db, ok := connPool.Get().(*sql.DB)
+	if !ok {
+		log.Fatal("The type of db getted from pool is wrong")
```
Comment (andelf) [inline sql-bench/main.go]: getted

File: sql-bench/main.go [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+		spend       time.Duration
+		tempStartTs = startTs
+		tempTotal   int64
+		tempSpend   time.Duration
+		tempSucc    int64
+	)
+	for {
+		tempExecTime := time.Now().Sub(tempStartTs)
+		if *reportInterval != 0 && tempExecTime.Seconds() >= float64(*reportInterval) {
+			fmt.Printf("Query: %d, Succ: %d, Faild: %d, Time: %v, Avg response time: %.04fms, QPS: %.02f : \n", tempTotal, tempSucc, tempTotal-tempSucc, tempExecTime, (tempSpend.Seconds()*1000)/float64(tempTotal), float64(tempTotal)/tempExecTime.Seconds())
```
Comment (andelf) [inline sql-bench/main.go]: use log

**Comment by andelf** [comment]
LGTM

**Pull request merged**